scalding-core/src/main/scala/com/twitter/scalding/typed/TypedPipe.scala

```diff
@@ -1131,6 +1132,28 @@ class MappablePipeJoinEnrichment[T](pipe: TypedPipe[T]) {
 }
 
 /**
+ * This class is to enrich the forceToDisk operation on TypedPipes and allow
+ * us to potentially optimize these operations when they're chained together
+ * redundantly.
+ */
+case class ForcedToDiskTypedPipe[T](typedPipe: TypedPipe[T]) extends TypedPipe[T] {
+  import Dsl._
+
+  override def forceToDisk: TypedPipe[T] = this
+
+  override def asPipe[U >: T](fieldNames: Fields)(implicit flowDef: FlowDef, mode: Mode, setter: TupleSetter[U]): Pipe = {
+    val pipe = typedPipe.toPipe[U](fieldNames)(flowDef, mode, setter)
+    pipe.forceToDisk
+  }
+
+  override def cross[U](tiny: TypedPipe[U]): TypedPipe[(T, U)] =
```

Comment:
see `TypedPipeFactory` for this use case then.

Do wish we had a nice ADT here.  :/